libs/game/sprite.ts

```diff
@@ -214,18 +215,103 @@ class Sprite implements SpriteLike {
     /**
      * Display a speech bubble with the text, for the given time
      * @param text the text to say, eg: ":)"
-     * @param time time to keep text on, eg: 2000
+     * @param time time to keep text on
      */
     //% group="Properties"
     //% blockId=spritesay block="%sprite(agent) say %text||for %millis ms"
-    //% time.defl=2000
+    //% inlineInputMode=inline
     //% help=sprites/sprite/say
-    say(text: string, millis?: number) {
-        this._say = text;
-        if (!millis || millis < 0)
-            this._sayExpires = -1;
-        else
-            this._sayExpires = control.millis() + millis;
+    say(text: string, timeOnScreen?: number, textColor = 15, textBoxColor = 1) {
+        if (!text) {
+            this.updateSay = undefined;
+            return;
+        }
+
+        if (timeOnScreen) {
+            timeOnScreen = timeOnScreen + control.millis();
+        }
+        let pixelsOffset = 0;
+        let holdTextTimer = 1.5;
+        let bubblePadding = 4;
+        let maxTextWidth = 100;
+        let font = image.font8;
+        let startX = 2;
+        let startY = 2;
+        let bubbleWidth = text.length * font.charWidth + bubblePadding;
+        let maxOffset = text.length * font.charWidth - maxTextWidth;
+        let spriteTop = this._hitboxes[0].top;
+
+        for (let i = 0; i < this._hitboxes.length; i++) {
+            if (this._hitboxes[i].top < spriteTop) {
+                spriteTop = this._hitboxes[i].top;
```

Comment:
The more JavaScript-y way to write this is:
`spriteTop = Math.min(spriteTop, this._hitboxes[i].top)`